File: Problems/Intervals/Non-overlapping-Intervals/Non-overlapping-Intervals.cs
using System;
using System.Collections.Generic;

// Using a greedy algorithm - Time: O(nlogn)

public class Solution
{
    public int EraseOverlapIntervals(int[][] intervals)
    {
        Array.Sort(intervals, (a, b) => a[1].CompareTo(b[1]));

        int ans = 0;
        int end = int.MinValue; // end 

        foreach (var i in intervals)
        {
            if (i[0] >= end)
                end = i[1];
            else
                ans++;
        }

        return ans;
    }
}

class Program
{
    static void Main(string[] args)
    {
        int[][] intervals = new int[][] {
            new int[] {1, 2},
            new int[] {2, 3},
            new int[] {3, 4},
            new int[] {1, 3}
        };
        
        Console.Write("Input: intervals = [");
        foreach (var i in intervals)
        {
            Console.Write("[" + i[0] + "," + i[1] + "]");
            if(i != intervals[intervals.Length - 1])
                Console.Write(",");
        }
        Console.WriteLine("]");

        Solution sol = new Solution();
        int ans = sol.EraseOverlapIntervals(intervals);

        Console.WriteLine("Output: " + ans);
    }
}

File: Problems/Intervals/Non-overlapping-Intervals/Non-overlapping-Intervals.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

// Using a greedy algorithm - Time: O(nlogn)

class Solution {
public:
    int eraseOverlapIntervals(std::vector<std::vector<int>> &intervals)
    {
        sort(begin(intervals), end(intervals), [](auto &a, auto &b) { return a[1] < b[1]; });

        int ans = 0;
        int end = INT_MIN;

        for (auto &i : intervals)
        {
            if (i[0] >= end)
                end = i[1];
            else
                ans++;
        }

        return ans;
    }
};

int main() 
{    
    std::vector<std::vector<int>> intervals = {{1, 2}, {2, 3}, {3, 4}, {1, 3}};

    std::cout << "Input: intervals = [";
    for (const auto &interval : intervals)
    {
        std::cout << "[" << interval[0] << "," << interval[1] << "]";
        if (&interval != &intervals.back())
            std::cout << ",";
    }
    std::cout << "]" << std::endl;
    
    Solution sol;
    int ans = sol.eraseOverlapIntervals(intervals);
    
    std::cout << "Output: " << ans << std::endl;
    
    return 0;
}